kroxylicious-sample/src/test/java/io/kroxylicious/sample/SampleProduceRequestFilterTest.java

```diff
@@ -0,0 +1,157 @@
+    private static final String POST_TRANSFORM_VALUE = "this is what the value will be transformed to";
+    private static final String CONFIG_FIND_VALUE = "from";
+    private static final String CONFIG_REPLACE_VALUE = "to";
+    private static final ApiMessageType API_MESSAGE_TYPE = ApiMessageType.PRODUCE;
+
+    @Mock
+    private KrpcFilterContext context;
+
+    @Captor
+    private ArgumentCaptor<ApiMessage> apiMessageCaptor = ArgumentCaptor.forClass(ApiMessage.class);
+
+    private FilterInvoker invoker;
+
+    @BeforeEach
+    public void beforeEach() {
+        buildContextMock();
+        SampleFilterConfig config = new SampleFilterConfig(CONFIG_FIND_VALUE, CONFIG_REPLACE_VALUE);
+        SampleProduceRequestFilter filter = new SampleProduceRequestFilter(config);
+        invoker = FilterInvokers.from(filter);
+    }
+
+    /**
+     * Unit Test: Checks that transformation is applied when request data contains configured value.
+     */
+    @Test
+    public void willTransformProduceRequestTest() {
+        var headerData = new RequestHeaderData();
+        var requestData = buildProduceRequestData(PRE_TRANSFORM_VALUE);
+        invoker.onRequest(ApiKeys.forId(API_MESSAGE_TYPE.apiKey()), API_VERSION, headerData, requestData, context);
+        verify(context).forwardRequest(any(), apiMessageCaptor.capture());
+        var unpackedRequest = unpackProduceRequestData((ProduceRequestData) apiMessageCaptor.getValue());
+        // We only put 1 record in, we should only get 1 record back, and
+        // We should see that the unpacked request value has changed from the input value, and
+        // We should see that the unpacked request value has been transformed to the correct value
+        assertThat(unpackedRequest)
+                .hasSize(1)
+                .first().asString()
+                .doesNotContain(PRE_TRANSFORM_VALUE)
+                .contains(POST_TRANSFORM_VALUE);
+    }
+
+    /**
+     * Unit Test: Checks that transformation is not applied when request data does not contain configured
+     * value.
+     */
+    @Test
+    public void wontTransformProduceRequestTest() {
+        var headerData = new RequestHeaderData();
+        var requestData = buildProduceRequestData(NO_TRANSFORM_VALUE);
+        invoker.onRequest(ApiKeys.forId(API_MESSAGE_TYPE.apiKey()), API_VERSION, headerData, requestData, context);
+        verify(context).forwardRequest(any(), apiMessageCaptor.capture());
+        var unpackedRequest = unpackProduceRequestData((ProduceRequestData) apiMessageCaptor.getValue());
+        // We only put 1 record in, we should only get 1 record back, and
+        // We should see that the unpacked request value has not changed from the input value
+        assertThat(unpackedRequest)
+                .hasSize(1)
+                .first().asString()
+                .contains(NO_TRANSFORM_VALUE);
+    }
+
+    private void buildContextMock() {
+        context = mock(KrpcFilterContext.class);
```

Comment:
For completeness this is doing the same as the Fetch test and replacing the annotated field.

Reply:
Removed the annotations (because using the annotations alone produced errors where context did not consistently appear instantiated). It's now just using the mock function.